bin/inx-chronicle/src/listener.rs

```diff
@@ -1,75 +1,198 @@
+
+        cx.spawn_actor_supervised::<InxStreamListener<inx::proto::Message>, _>(
+            InxStreamListener::new(self.broker_handle.clone())?.with_stream(message_stream),
+        )
+        .await;
+        cx.spawn_actor_supervised::<InxStreamListener<inx::proto::Milestone>, _>(
+            InxStreamListener::new(self.broker_handle.clone())?.with_stream(milestone_stream),
+        )
+        .await;
+        Ok(inx_client)
+    }
+}
+
+#[async_trait]
+impl HandleEvent<Report<InxStreamListener<inx::proto::Message>>> for InxListener {
+    async fn handle_event(
+        &mut self,
+        cx: &mut ActorContext<Self>,
+        event: Report<InxStreamListener<inx::proto::Message>>,
+        inx_client: &mut Self::Data,
+    ) -> Result<(), Self::Error> {
+        // TODO: Figure out why `cx.shutdown()` is not working.
+        let handle = cx.handle();
+        match event {
+            Ok(_) => {
+                handle.shutdown().await;
+            }
+            Err(e) => match e.error {
+                ActorError::Result(_) | ActorError::Panic => {
+                    let message_stream = inx_client.listen_to_messages(MessageFilter {}).await?.into_inner();
+                    cx.spawn_actor_supervised::<InxStreamListener<inx::proto::Message>, _>(
+                        InxStreamListener::new(self.broker_handle.clone())?.with_stream(message_stream),
+                    )
+                    .await;
+                }
+                ActorError::Aborted => {
+                    handle.shutdown().await;
+                }
+            },
+        }
+        Ok(())
+    }
+}
+
+#[async_trait]
+impl HandleEvent<Report<InxStreamListener<inx::proto::Milestone>>> for InxListener {
+    async fn handle_event(
+        &mut self,
+        cx: &mut ActorContext<Self>,
+        event: Report<InxStreamListener<inx::proto::Milestone>>,
+        inx_client: &mut Self::Data,
+    ) -> Result<(), Self::Error> {
+        // TODO: Figure out why `cx.shutdown()` is not working.
+        let handle = cx.handle();
+        match event {
+            Ok(_) => {
+                handle.shutdown().await;
+            }
+            Err(e) => match e.error {
+                ActorError::Result(_) | ActorError::Panic => {
+                    let milestone_stream = inx_client.listen_to_latest_milestone(NoParams {}).await?.into_inner();
+                    cx.spawn_actor_supervised::<InxStreamListener<inx::proto::Milestone>, _>(
+                        InxStreamListener::new(self.broker_handle.clone())?.with_stream(milestone_stream),
+                    )
+                    .await;
+                }
+                ActorError::Aborted => {
+                    handle.shutdown().await;
+                }
+            },
         }
+        Ok(())
     }
+}
+
+pub struct InxStreamListener<I> {
+    broker_handle: Addr<Broker>,
+    _item: PhantomData<I>,
```

Comment:
Probably too nit-picky and honestly I don't care, but maybe `_marker` or `_phantom` is a better field name? At least those are the names I see most often. This comment is more to make sure you chose this name deliberately. Close if you do! :slightly_smiling_face:

Reply:
I chose it simply because it represents the stream's `Item`.